x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrailTests.java

```diff
@@ -279,6 +343,738 @@ public void testAuditFilterSettingValidation() {
         assertThat(e, hasToString(containsString("invalid pattern [/no-inspiration]")));
     }
 
+    public void testSecurityConfigChangeEventFormattingForRoles() throws IOException {
+        final Path path = getDataPath("/org/elasticsearch/xpack/security/audit/logfile/audited_roles.txt");
+        final Map<String, String> auditedRolesMap = new HashMap<>();
+        try (BufferedReader reader = new BufferedReader(new InputStreamReader(new BufferedInputStream(Files.newInputStream(path)),
+                StandardCharsets.UTF_8))) {
+            String line;
+            while ((line = reader.readLine()) != null) {
+                // even number of lines
+                auditedRolesMap.put(line, reader.readLine());
+            }
+        }
+
+        RoleDescriptor nullRoleDescriptor = new RoleDescriptor("null_role", randomFrom((String[]) null, new String[0]),
+                randomFrom((RoleDescriptor.IndicesPrivileges[]) null, new RoleDescriptor.IndicesPrivileges[0]),
+                randomFrom((RoleDescriptor.ApplicationResourcePrivileges[])null, new RoleDescriptor.ApplicationResourcePrivileges[0]),
+                randomFrom((ConfigurableClusterPrivilege[])null, new ConfigurableClusterPrivilege[0]),
+                randomFrom((String[])null, new String[0]),
+                randomFrom((Map<String, Object>)null, Map.of()),
+                Map.of("transient", "meta", "is", "ignored"));
```

Comment:
I don't understand what's wrong with asserting the audit format for an empty role.

Reply:
First this should have been a Nit. Second I think we talked about different things. It's not about the empty role itself. It is about the randomised arguments used to construct it and I think the randomisation is not necessary if we have a test for `RoleDescriptor` itself looks like the follows:
```java
var name = randomAlphaOfLengthBetween(1, 10);
var d1 = new RoleDescriptor(name, null, null, null, null, null, null, null);
var d2 = new RoleDescriptor(name, 
    new String[0], 
    new RoleDescriptor.IndicesPrivileges[0], 
    new RoleDescriptor.ApplicationResourcePrivileges[0], 
    new ConfigurableClusterPrivilege[0], 
    new String[0], Map.of(), Map.of());
assertTrue(d1.equals(d2));
```

The above would ensure that `null` and empty arrays work identically for `RoleDescriptor` and how they are managed internally is an implementation detail. Therefore its callers do not have to worry about getting different behaviour whether the passed-in arguments are null or empty array, so that tests don't have to random from them. Otherwise, each place where `RoleDescriptor` is used, the tests need to random between null and empty array, which are just duplicated code.

Reply:
I have pushed  8b3dcf0aa1684d0662ddfb95934c3f5 that adds a test as you're describing.
But I personally resist assuming that if two object are equals then they must audit to the same thing. I don't see the audit method as a serialization type of method.

If you insist, I will remove the randomization.

Reply:
It's OK to keep the randomization.

No code related actions are required, but I'd like to voice a bit more of my opinion on this topic: 

We can in general assume if two objects are equal, they should beheave the same regardless of actions (audit included). After all, that's what `equals` is mostly used for. For an example, when reloading role definitions from a file, our code will cleverly skip the update if the newly parsed `RoleDescriptor` is *equal* to the existing one. The message here is that if two objects are equal, they are interchangable in terms of bebaviours and auditing should be no exception. The`equals` method is not about just serialization, sure two equal objects should serialise the same as well, but `equals` should guarantee more than just that. Plus we control the code of `RoleDescriptor`, so it is reasonable to expect it is well implemented to provide a reliable `equals` method. Unless what we care about is object identity, but that is *not* something we are talking about here.